File: src/module/models/bangumi.py
from pydantic import BaseModel, Field
from dataclasses import dataclass


class BangumiData(BaseModel):
    id: int = Field(0, alias="id", title="番剧ID")
    official_title: str = Field("official_title", alias="official_title", title="番剧中文名")
    year: str | None = Field(None, alias="year", title="番剧年份")
    title_raw: str = Field("title_raw", alias="title_raw", title="番剧原名")
    season: int = Field(1, alias="season", title="番剧季度")
    season_raw: str | None = Field(None, alias="season_raw", title="番剧季度原名")
    group_name: str | None = Field(None, alias="group_name", title="字幕组")
    dpi: str | None = Field(None, alias="dpi", title="分辨率")
    source: str | None = Field(None, alias="source", title="来源")
    subtitle: str | None = Field(None, alias="subtitle", title="字幕")
    eps_collect: bool = Field(False, alias="eps_collect", title="是否已收集")
    offset: int = Field(0, alias="offset", title="番剧偏移量")
    filter: list[str] = Field(["720", "\\d+-\\d+"], alias="filter", title="番剧过滤器")
    rss_link: list[str] = Field([], alias="rss_link", title="番剧RSS链接")
    poster_link: str | None = Field(None, alias="poster_link", title="番剧海报链接")
    added: bool = Field(False, alias="added", title="是否已添加")
    rule_name: str | None = Field(None, alias="rule_name", title="番剧规则名")
    save_path: str | None = Field(None, alias="save_path", title="番剧保存路径")
    deleted: bool = Field(False, alias="deleted", title="是否已删除")


class Notification(BaseModel):
    official_title: str = Field(..., alias="official_title", title="番剧名")
    season: int = Field(..., alias="season", title="番剧季度")
    episode: int = Field(..., alias="episode", title="番剧集数")


@dataclass
class Episode:
    title_en: str | None
    title_zh: str | None
    title_jp: str | None
    season: int
    season_raw: str
    episode: int
    sub: str
    group: str
    resolution: str
    source: str


@dataclass
class SeasonInfo(dict):
    official_title: str
    title_raw: str
    season: int
    season_raw: str
    group: str
    filter: list | None
    offset: int | None
    dpi: str
    source: str
    subtitle: str
    added: bool
    eps_collect: bool

File: webui/types/bangumi.ts
/**
 * @type `Bangumi` in backend/src/module/models/bangumi.py
 */
export interface BangumiRule {
  added: boolean;
  deleted: boolean;
  archived: boolean;
  dpi: string;
  eps_collect: boolean;
  filter: string[];
  group_name: string;
  id: number;
  official_title: string;
  episode_offset: number;
  season_offset: number;
  poster_link: string | null;
  rss_link: string[];
  rule_name: string;
  save_path: string;
  season: number;
  season_raw: string;
  source: string | null;
  subtitle: string;
  title_raw: string;
  year: string | null;
  air_weekday: number | null; // 0=Mon, 1=Tue, ..., 6=Sun, null=Unknown
  needs_review: boolean;
  needs_review_reason: string | null;
}

export interface BangumiAPI extends Omit<BangumiRule, 'filter' | 'rss_link'> {
  filter: string;
  rss_link: string;
}

export interface SearchResult {
  order: number;
  value: BangumiRule;
}

export type BangumiUpdate = Omit<BangumiAPI, 'id'>;

export const ruleTemplate: BangumiRule = {
  added: false,
  deleted: false,
  archived: false,
  dpi: '',
  eps_collect: false,
  filter: [],
  group_name: '',
  id: 0,
  official_title: '',
  episode_offset: 0,
  season_offset: 0,
  poster_link: '',
  rss_link: [],
  rule_name: '',
  save_path: '',
  season: 1,
  season_raw: '',
  source: null,
  subtitle: '',
  title_raw: '',
  year: null,
  air_weekday: null,
  needs_review: false,
  needs_review_reason: null,
};

/** Legacy offset suggestion (for backward compatibility) */
export interface OffsetSuggestion {
  suggested_offset: number;
  reason: string;
}

/** TMDB summary for display in offset dialog */
export interface TMDBSummary {
  title: string;
  total_seasons: number;
  season_episode_counts: Record<number, number>;
  status: string | null;
}

/** Detailed offset suggestion from detector */
export interface OffsetSuggestionDetail {
  season_offset: number;
  episode_offset: number;
  reason: string;
  confidence: 'high' | 'medium' | 'low';
}

/** Request for detect-offset API */
export interface DetectOffsetRequest {
  title: string;
  parsed_season: number;
  parsed_episode: number;
}

/** Response from detect-offset API */
export interface DetectOffsetResponse {
  has_mismatch: boolean;
  suggestion: OffsetSuggestionDetail | null;
  tmdb_info: TMDBSummary | null;
}
